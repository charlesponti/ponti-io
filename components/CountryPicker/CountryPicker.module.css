.container {
  display: flex;
  justify-content: center;
  margin: 16px 0;
  max-width: 700px;
}

.container label {
  flex-grow: 0;
  flex-basis: 91.666667%;
}

.form-control {
  min-width: 100% !important;
}

.select {
  background-color: #131313;
  border-radius: 4px;
  display: block;
  color: white;
  font-size: 16px;
  margin-top: 8px;
  padding: 10px;
  width: 100%;

  /* reset */
  -webkit-box-sizing: border-box;
  -moz-box-sizing: border-box;
  box-sizing: border-box;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;

  /* arrow */
  background-image: linear-gradient(45deg, transparent 50%, gray 50%),
    linear-gradient(135deg, gray 50%, transparent 50%),
    linear-gradient(to right, #ccc, #ccc);
  background-position:
    calc(100% - 20px) calc(1em + 2px),
    calc(100% - 15px) calc(1em + 2px),
    calc(100% - 2.5em) 0.5em;
  background-size:
    5px 5px,
    5px 5px,
    1px 1.5em;
  background-repeat: no-repeat;
}
