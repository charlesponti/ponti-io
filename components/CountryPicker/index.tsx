import type { ChangeEvent } from "react";
import { useCallback } from "react";
import styled from "styled-components";

import FeedbackBlock from "@/components/FeedbackBlock";
import Grid from "@/components/Grid";
import { BASE_URL, type CountryData } from "@/utils/corona.api";

import { useQuery } from "@tanstack/react-query";
import styles from "./CountryPicker.module.css";

const Label = styled.label`
  display: block;
  font-weight: 600;
  width: 100%;
`;

type CountryPickerProps = {
	readonly onChange: (val: string) => void;
	readonly countryCode: string;
};
function CountryPicker({
	onChange,
	countryCode,
}: CountryPickerProps): JSX.Element {
	const {
		data: countries,
		isLoading,
		isError,
	} = useQuery<CountryData[]>({
		queryKey: ["countries"],
		queryFn: () => fetch(`${BASE_URL}/countries`).then((res) => res.json()),
	});

	const onSelectChange = useCallback(
		(event: ChangeEvent<HTMLSelectElement>) => {
			onChange(event.target.value);
		},
		[onChange],
	);

	if (isLoading) {
		return <div>Loading countries...</div>;
	}

	if (isError || !countries) {
		return (
			<FeedbackBlock>
				<p>Unable to load countries</p>
			</FeedbackBlock>
		);
	}

	return (
		<Grid className={styles.container}>
			<Label htmlFor="country" data-testid="country-picker">
				Country
				<select
					id="country"
					className={styles.select}
					onChange={onSelectChange}
					value={countryCode}
				>
					<option value="global">All</option>
					{countries.map((country) => (
						<option key={country.name} value={country.iso3}>
							{country.name}
						</option>
					))}
				</select>
			</Label>
		</Grid>
	);
}

export default CountryPicker;
