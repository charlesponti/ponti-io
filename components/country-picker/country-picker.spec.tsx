import renderWithProviders from "@/tests/test.utils";
import { screen, waitFor } from "@testing-library/react";
import { type Mock, beforeEach, describe, expect, test, vi } from "vitest";

// Import the server from the setup file instead of creating a new one
import CountryPicker from "./country-picker";

describe("CountryPicker", () => {
	let onCountryChange: Mock;

	// No need for server setup/teardown here since it's handled in test.setup.ts

	beforeEach(() => {
		onCountryChange = vi.fn();
	});

	test("should render country options", async () => {
		const { findByText } = renderWithProviders(
			<CountryPicker countryCode="US" onChange={onCountryChange} />,
		);

		// Wait for loading to finish and component to render
		await waitFor(() => screen.queryByTestId("country-picker"));

		// Verify countries are displayed
		expect(await findByText("United States")).toBeTruthy();
		expect(await findByText("Canada")).toBeTruthy();
		expect(await findByText("Italy")).toBeTruthy();
		expect(await findByText("China")).toBeTruthy();
	});

	test("should render global data", async () => {
		const { container } = renderWithProviders(
			<CountryPicker countryCode="global" onChange={onCountryChange} />,
		);

		await waitFor(() => screen.queryByTestId("country-picker"));
		expect(container).toMatchSnapshot();
	});

	test("should render US data", async () => {
		const { container } = renderWithProviders(
			<CountryPicker countryCode="US" onChange={onCountryChange} />,
		);

		await waitFor(() => screen.queryByTestId("country-picker"));
		expect(container).toMatchSnapshot();
	});
});
