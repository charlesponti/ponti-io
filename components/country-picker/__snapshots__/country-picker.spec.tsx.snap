// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CountryPicker > should render US data 1`] = `
<div>
  <div
    class="_container_ef5ebd"
  >
    <label
      class="sc-gjcoXW dAhsdb"
      data-testid="country-picker"
      for="country"
    >
      Country
      <select
        class="_select_ef5ebd"
        id="country"
      >
        <option
          value="global"
        >
          All
        </option>
        <option
          value="USA"
        >
          United States
        </option>
        <option
          value="CAN"
        >
          Canada
        </option>
        <option
          value="CHN"
        >
          China
        </option>
        <option
          value="ITA"
        >
          Italy
        </option>
        <option
          value="GLB"
        >
          Global
        </option>
      </select>
    </label>
  </div>
</div>
`;

exports[`CountryPicker > should render global data 1`] = `
<div>
  <div
    class="_container_ef5ebd"
  >
    <label
      class="sc-gjcoXW dAhsdb"
      data-testid="country-picker"
      for="country"
    >
      Country
      <select
        class="_select_ef5ebd"
        id="country"
      >
        <option
          value="global"
        >
          All
        </option>
        <option
          value="USA"
        >
          United States
        </option>
        <option
          value="CAN"
        >
          Canada
        </option>
        <option
          value="CHN"
        >
          China
        </option>
        <option
          value="ITA"
        >
          Italy
        </option>
        <option
          value="GLB"
        >
          Global
        </option>
      </select>
    </label>
  </div>
</div>
`;
